### ISSUE TITLE
Allow importing librosa on Python 2 without a warning

### ISSUE THREAD

**Issue by bluenote10** [librosa/librosa#990]
#### Description

I'd like to use librosa in Python 2 without an annoying warning on every import. The common pattern in Python is:

#### Steps/Code to Reproduce

```python
import warnings

with warnings.catch_warnings():
    warnings.filterwarnings("ignore", category=DeprecationWarning)
    import librosa
```
Unfortunately that doesn't work, because librosa resets the filterwarning itself on import, which is an antipattern.

#### Expected Results

No warning on import.

#### Actual Results

3 lines (wrapped) of stdout noise:
```
/home/fabian/.virtualenvs/default/local/lib/python2.7/site-packages/librosa/__init__.py:40: DeprecationWarning: You are using librosa with Python 2. Please note that librosa 0.7 will be the last version to support Python 2, after which it will require Python 3 or later.
  DeprecationWarning)
```

#### Versions

```
Linux-4.15.0-62-generic-x86_64-with-Ubuntu-18.04-bionic
('Python', '2.7.15rc1 (default, Apr 15 2018, 21:51:34) \n[GCC 7.3.0]')
('NumPy', '1.15.2')
('SciPy', '1.1.0')
/home/fabian/.virtualenvs/default/local/lib/python2.7/site-packages/librosa/__init__.py:40: DeprecationWarning: You are using librosa with Python 2. Please note that librosa 0.7 will be the last version to support Python 2, after which it will require Python 3 or later.
  DeprecationWarning)
('librosa', '0.7.0')
```

**Comment by bmcfee** [comment]
> Unfortunately that doesn't work, because librosa resets the filterwarning itself on import, which is an antipattern.

I agree that it's not a great practice.  As I recall, the reason for this was to get around ipython (and similar tools) doing this kind of thing, which was hiding warnings that really should have been bubbled up to the user level (eg function renames and deprecations).  We reasoned that resetting this filter on import was the only way to achieve that functionality, and that it was worth compromising on style to make sure that warnings are not accidentally hidden from view.

Perhaps the situation in ipython has changed in the intervening years?  If not, I'm extremely hesitant to remove this reset, even if it is an antipattern, because it ultimately does more good than harm.

**Comment by bluenote10** [comment]
> As I recall, the reason for this was to get around ipython (and similar tools) doing this kind of thing, which was hiding warnings that really should have been bubbled up to the user level (eg function renames and deprecations).

That's an issue of these tools than. I have never seen that behavior in IPython or any other tool to be honest. Client code should be in control of warning behavior. Depriving clients of control just because of some broken tool is a bit sad, and solves the problem by copying the bad practice of that tool just in the other direction.

Incidentally, a simple solution to silent this particular warning would be to swap the order of re-enabling the filterwarning and the Python 2 check. But I still think hard coding warnings to `always` on import should be avoided in general.

**Comment by bmcfee** [comment]
> That's an issue of these tools than. I have never seen that behavior in IPython or any other tool to be honest. Client code should be in control of warning behavior. Depriving clients of control just because of some broken tool is a bit sad, and solves the problem by copying the bad practice of that tool just in the other direction.

I agree, but I'm also sympathetic to practical solutions.  That was the practical solution at the time.

That said, this does appear to have been addressed here: https://github.com/ipython/ipython/issues/8478/ , but I'd like to verify that things work properly before committing to changing our current behavior.

**Comment by bluenote10** [comment]
Okay that issue has been closed over 4 years ago, which is probably why I never faced that issue. For what it's worth, I'm seeing deprecation warnings in IPython all the time:

![image](https://user-images.githubusercontent.com/3620703/65283058-a4233b00-db36-11e9-9b85-58c24ceac553.png)

**Comment by bmcfee** [comment]
> Okay that issue has been closed over 4 years ago, which is probably why I never faced that issue. For what it's worth, I'm seeing deprecation warnings in IPython all the time:

Interesting.  I just spun up a fresh python 2.7 conda environment, and disabled the warnings reset on my local branch of librosa.  The warning is still suppressed.

```
 🐍 legacy  ~ 
 →  ipython
Python 2.7.16 |Anaconda, Inc.| (default, Aug 22 2019, 16:00:36) 
Type "copyright", "credits" or "license" for more information.

IPython 5.8.0 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: import librosa

In [2]: 
```

This also happens with a completely fresh profile (`--init`), so it shouldn't be down to any old configuration data being loaded by the new environment.  It seems like this is not actually resolved yet, but I'll look into it.

**Comment by bmcfee** [comment]
>  It seems like this is not actually resolved yet, but I'll look into it.

I've done a bit more digging, trying it out on a separate machine, and I still get the same behavior (deprecation warnings do not show by default).  Until we can track down the source of this problem and understand it completely, I don't think we should remove the warning filter reset #991 .

**Comment by bluenote10** [comment]
We should at least swap the order of issuing the Python 2 warning and re-enabling the filterwarnings. 

Reenabling first and warning immediately after gives no option of Python 2 user to get nice stdout out of any piece of code that uses librosa. Doing it the other way around at least enables clients to do something about that warning coming out of librosa with no downsides (and with context managers they could also undo that filterwarnings changes librosa makes).

I have updated to PR to do that.

**Comment by bmcfee** [comment]
> We should at least swap the order of issuing the Python 2 warning and re-enabling the filterwarnings.

I disagree.  The entire point of resetting the filter is to surface deprecation warnings.  If we reset after, we may as well not reset at all.



>  Python 2 user to get nice stdout out of any piece of code that uses librosa

Warnings go to stderr, not stdout, so you should be able to redirect them to avoid polluting the log.

**Comment by bluenote10** [comment]
> I disagree. The entire point of resetting the filter is to surface deprecation warnings. If we reset after, we may as well not reset at all.

Okay, I though you were more worried about deprecation warnings coming from some other library that librosa uses.

I'm not sure if the value of that message justifies enforcing it. Basically what it states is most likely true for many other libraries X: There will be a future version of library X that drops Python 2 support. However, version 0.7 works perfectly fine in Python 2, so why should it issue an unconditional warning? A warning should be something that can be resolved. This raises the question if Python 2 users are better off downgrading to an earlier version of librosa that doesn't warn. Or what if the next version of librosa happens to still work with Python 2?

In my opinion, warnings are simply not the right communication channel for future release information.

**Comment by bmcfee** [comment]
> Okay, I though you were more worried about deprecation warnings coming from some other library that librosa uses.

That too, but that would be our job to fix.  Librosa itself deprecates functionality from time to time, and I want to make sure that users are aware when that happens and have sufficient time to revise their code to avoid breakage.  Our typical deprecation cycle is about 1 year, but that only really works if the warnings are issued in such a way as the users can see them.



> However, version 0.7 works perfectly fine in Python 2, so why should it issue an unconditional warning? A warning should be something that can be resolved.

Yes, that's why it's a warning and not an error / installation failure.  The resolution to this warning is to move to python 3.




>  Or what if the next version of librosa happens to still work with Python 2?

It most likely will not.  We have a handful of shims in place to preserve functionality across major versions, but I do not want to maintain these indefinitely, and they will be removed in 0.8.  The installation requirements will be revised as well.  Of course, users can always do whatever they wish with the code (including hacking it to run on py2) but we make no claims going forward as to whether that will work or not.



> In my opinion, warnings are simply not the right communication channel for future release information.

Do you have an alternative mechanism in mind?  Deprecation warnings are commonly used in (scientific) python libraries, and I can't think of any other mechanism that would work as well.

Just to clarify my position on everything, in case it got lost somewhere along the way:

1. We wlil not support python 2 from version 0.8 onward.
2. Resetting the warnings filter is an ugly hack to work around a quirk in ipython, and I agree it should be removed, so long as our desired behavior is preserved.
3. I thought that the ipython quirk had been resolved, but it does not appear to be on my system(s).  Independent verification of this would be very much appreciated.

**Comment by bluenote10** [comment]
Is it possible that it is an IPython user setting? For example, having some configuration files in `~/.ipython` or `~/.config/ipython` (see [IPython configration docs](https://ipython.readthedocs.io/en/stable/development/config.html)) that does something like this [`~/.ipython/profile_default/startup/disable-warnings.py`](https://stackoverflow.com/a/9031848/1804173) example? I've really never heard of this behavior and can't reproduce it.

**Comment by bmcfee** [comment]
It's possible, but I've checked for those things on my local installations and haven't found anything suggestive.  I haven't had the time to properly dig into it on a completely fresh installation though.

**Comment by bmcfee** [comment]
I did a bit more digging, and I think we can rule out ipython as the culprit -- vanilla python console has the same behavior.

According to the python 3.7 [docs](https://docs.python.org/3/library/warnings.html#warning-categories) this is actually expected behavior (?!); deprecation warnings are hidden *unless* triggered by code called from a `__main__()` function.

However, the docs are a little different in [3.6](https://docs.python.org/3.6/library/warnings.html#warning-categories), where there is no mention of the `__main__` exception.  2.7 also ignores deprecation warnings by default.

# Solution

If we change the 2.7 deprecation over to a `FutureWarning` (which is arguably a better fit for this purpose, though we're not targetting end-user applications here), I think we can drop the filter and keep all the desired behavior.  @bluenote10 does that sound agreeable to you?

**Comment by bluenote10** [comment]
Yes sounds good to me, thanks for taking the time to look into this!
